exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -1073,6 +494,617 @@ func (b *Bitfinex) wsHandleData(respRaw []byte) error {
+		}
+		tradeHolder = append(tradeHolder, wsTrade)
+	}
+	trades := make([]trade.Data, len(tradeHolder))
+	for i := range tradeHolder {
+		side := order.Buy
+		newAmount := tradeHolder[i].Amount
+		if newAmount < 0 {
+			side = order.Sell
+			newAmount *= -1
+		}
+		price := tradeHolder[i].Price
+		if price == 0 && tradeHolder[i].Rate > 0 {
+			price = tradeHolder[i].Rate
+		}
+		trades[i] = trade.Data{
+			TID:          strconv.FormatInt(tradeHolder[i].ID, 10),
+			CurrencyPair: c.Currency,
+			Timestamp:    time.UnixMilli(tradeHolder[i].Timestamp),
+			Price:        price,
+			Amount:       newAmount,
+			Exchange:     b.Name,
+			AssetType:    c.Asset,
+			Side:         side,
+		}
+	}
+
+	return b.AddTradesToBuffer(trades...)
+}
+
+func (b *Bitfinex) handleWSNotification(d []interface{}, respRaw []byte) error {
+	notification, ok := d[2].([]interface{})
+	if !ok {
+		return errors.New("unable to type assert notification data")
+	}
+	if data, ok := notification[4].([]interface{}); ok {
+		channelName, ok := notification[1].(string)
+		if !ok {
+			return errors.New("unable to type assert channelName")
+		}
+		switch {
+		case strings.Contains(channelName, wsFundingOfferNewRequest),
+			strings.Contains(channelName, wsFundingOfferUpdateRequest),
+			strings.Contains(channelName, wsFundingOfferCancelRequest):
+			if data[0] != nil {
+				if id, ok := data[0].(float64); ok && id > 0 {
+					if b.Websocket.Match.IncomingWithData(int64(id), respRaw) {
+						return nil
+					}
+					offer, err := wsHandleFundingOffer(data, true /* include rate real */)
+					if err != nil {
+						return err
+					}
+					b.Websocket.DataHandler <- offer
+				}
+			}
+		case strings.Contains(channelName, wsOrderNewRequest):
+			if data[2] != nil {
+				if cid, ok := data[2].(float64); ok && cid > 0 {
+					if b.Websocket.Match.IncomingWithData(int64(cid), respRaw) {
+						return nil
+					}
+					b.wsHandleOrder(data)
+				}
+			}
+		case strings.Contains(channelName, wsOrderUpdateRequest),
+			strings.Contains(channelName, wsOrderCancelRequest):
+			if data[0] != nil {
+				if id, ok := data[0].(float64); ok && id > 0 {
+					if b.Websocket.Match.IncomingWithData(int64(id), respRaw) {
+						return nil
+					}
+					b.wsHandleOrder(data)
+				}
+			}
+		default:
+			return fmt.Errorf("%s - Unexpected data returned %s",
+				b.Name,
+				respRaw)
+		}
+	}
+	if notification[5] != nil {
+		if wsErr, ok := notification[5].(string); ok {
+			if strings.EqualFold(wsErr, wsError) {
+				if errMsg, ok := notification[6].(string); ok {
+					return fmt.Errorf("%s - Error %s",
+						b.Name,
+						errMsg)
+				}
+				return fmt.Errorf("%s - unhandled error message: %v", b.Name,
+					notification[6])
+			}
+		}
+	}
+	return nil
+}
+
+func (b *Bitfinex) handleWSPositionSnapshot(d []interface{}) error {
+	snapBundle, ok := d[2].([]interface{})
+	if !ok || len(snapBundle) == 0 {
+		return nil
+	}
+	if _, ok := snapBundle[0].([]interface{}); !ok {
+		return nil
```

Comment:
return error

Reply:
Actually ... I have no idea why this is here. Deleting it.